Reworked our content around Enterprise alerts (#224)
astronomer/docs: enterprise/v0.23/05_customize-airflow/04_airflow-alerts.md

## Overview

Whether you're just starting to use Apache Airflow or your team is running it at scale, incorporating an alerting framework is critical to the health of your data. On Astronomer Enterprise, users can leverage three types of alerting solutions:

- Airflow task and DAG-level alerts notify you via email when an Airflow task or DAG fails, succeeds, or retries. These require an SMTP (Simple Mail Transfer Protocol) service.

> - Deployment-level alerts notify you when the health of an Airflow Deployment is low or if any of Airflow's underlying components are underperforming, including the Airflow Scheduler.

Reviewer: ```suggestion
- Deployment-level alerts notify you when the health of your Airflow Deployment on Astronomer is low or if any of Airflow's underlying components are underperforming, including the Airflow Scheduler.
```

Author: I think "On Astronomer Enterprise" is carried over from the intro sentence into these bullets - including it again here seems redundant. Also since this is Enterprise, it's likely that readers will have more than one Deployment to manage which is why I went with "an" over "your".

Reviewer: Agreed with this now looking back at it!

> - Airflow task and DAG-level alerts notify you via email when an Airflow task or DAG fails, succeeds, or retries.
> - Deployment-level alerts notify you when the health of an Airflow Deployment is low or if any of Airflow's underlying components are underperforming, including the Airflow Scheduler.
> - Platform-level alerts notify you when a component of your Astronomer platform is unhealthy, such as Elasticsearch, Astronomer's Houston API, or your Docker Registry.

Reviewer: ```suggestion
- Airflow task and DAG-level alerts that notify you via email when an Airflow task or DAG fails, succeeds, or retries.
- Deployment-level alerts that notify you when the health of an Airflow Deployment is low or if any of Airflow's underlying components are underperforming, including the Airflow Scheduler.
- Platform-level alerts that notify you when a component of your Astronomer platform is unhealthy, such as Elasticsearch, Astronomer's Houston API, or your Docker Registry.
```

For some reason the transition here kept sounding off to me, maybe because we introduce these bullet points as "solutions" - and I'm reading each type of alert, and want to describe it as such - "task level alerts _that_ do x". Does that make sense?

Author: Changing this to `, which` instead of `that`, so that it's clear the focus is on the alerts rather than something that comes later in the process.

> This guide focuses on configuring task and DAG-level alerts. For information on configuring platform and Deployment alerts, read [Alerting in Astronomer Enterprise](https://www.astronomer.io/docs/enterprise/v0.23/monitor/platform-alerts).

Reviewer: ```suggestion
This guide focuses on configuring task and DAG-level alerts. For information on configuring platform and Deployment alerts, read [Alerting on Astronomer Enterprise](https://www.astronomer.io/docs/enterprise/v0.23/monitor/platform-alerts).
```

Alerting on?

- [SendGrid](https://sendgrid.com/)
- [Amazon SES](https://aws.amazon.com/ses/)

Step-by-step instructions on how to integrate these two services with Astronomer are provided below, but you can use any SMTP service for this purpose.

> **Note:** By default, email alerts for process failures are sent whenever individual tasks fail. To receive only 1 email per DAG failure, refer to the Limit Alerts to the DAG Level topic below. For more information and best practices on Airflow alerts, read [Error Notifications in Airflow](https://www.astronomer.io/guides/error-notifications-in-airflow/).

### Integrate SendGrid with Astronomer

[SendGrid](https://sendgrid.com/) is an email delivery service that's easy to set up for Airflow task-level alerts. A free SendGrid account grants users 40,000 free emails within the first 30 days of an account opening and 100 emails per day after that. This should be more than enough emails for most alerting use cases.

To get started with SendGrid:

> 1. [Create a SendGrid account](https://signup.sendgrid.com). When you create an account, be prepared to disclose some standard information about yourself and your organization.

Reviewer: ```suggestion
1. [Create a SendGrid account](https://signup.sendgrid.com). Be prepared to disclose some standard information about yourself and your organization.
```

To get started with SendGrid:

1. [Create a SendGrid Account](https://signup.sendgrid.com). When you create an account, be prepared to disclose some standard information about yourself and your organization.

2. [Verify a Single Sender Identity](https://sendgrid.com/docs/ui/sending-email/sender-verification/). Because you're sending emails only for internal administrative purposes, a single sender identity is sufficient for integrating with Astronomer. The email address you verify here is used as the sender for your Airflow alert emails.

3. Create a Key Using SendGrid's Web API. In SendGrid, go to **Email API** > **Integration Guide**. Follow the steps to generate a new API key using SendGrid's Web API and cURL.

4. Skip the step for exporting your API Key to your development environment. Instead, execute the generated curl code directly in your command line, making sure to replace `$SENDGRID_API_KEY` in the `--header` field with your copied key.

5. Verify your integration in SendGrid to confirm that the key was activated. If you get an error indicating that SendGrid can't find the test email, try rerunning the cURL code in your terminal before retrying the verification.

> 6. Open to your Airflow Deployment in the Astronomer UI, then go to the **Variables** tab and add the following Environment Variables using the **+Add** button:

Reviewer 2: ```suggestion
6. Open your Airflow Deployment in the Astronomer UI, then go to the **Variables** tab and add the following Environment Variables using the **+Add** button:
```

1. In the AWS Management Console, go to **AWS Console** > **Simple Email Service** > **Email Addresses** to add and verify the email addresses you want to receive alerts to.

2. Open the inbox of each email address you specified and verify them through the emails sent by Amazon.

3. In the AWS Console, go to **Simple Email Service** > **SMTP Settings** and use the **Create My SMTP Credentials** button to generate a username and password. This will look similar to an access and secret access key. Write down this username and password for later, as well as the **Server Name** and **Port**.

   > **Note:** You won't be able to access these values again, so consider storing them in a password manager.

4. Choose an Amazon EC2 region to use, then write down the code of this server for the next step. Refer to [Amazon's list of available regions and servers](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-regions) to determine which server best fits your needs.

5. Open your Airflow Deployment in the Astronomer UI and go to the **Variables** tab, then add the following Environment Variables using the **+Add** button:

   ```

>    AIRFLOW__SMTP__SMTP_HOST={Your SMTP host}

Reviewer 2: This is a bit nitpicky and up to you, but I prefer to format example variables like `<example-var>` vs `{Example var}`. I think they were already written in this format here. Just something to think about as you rework docs across the site.

Author: Ah yeah, good call. This was from the original doc and something I hadn't considered updating.

To get started with SendGrid:

> 1. [Create a SendGrid Account](https://signup.sendgrid.com). When you create an account, be prepared to disclose some standard information about yourself and your organization.

Reviewer: ```suggestion
1. [Create a SendGrid account](https://signup.sendgrid.com). When you create an account, be prepared to disclose some standard information about yourself and your organization.
```

Should these not be capitalized now that they're not headers?

2. [Verify a Single Sender Identity](https://sendgrid.com/docs/ui/sending-email/sender-verification/). Because you're sending emails only for internal administrative purposes, a single sender identity is sufficient for integrating with Astronomer. The email address you verify here is used as the sender for your Airflow alert emails.

> 3. Create a Key Using SendGrid's Web API. In SendGrid, go to **Email API** > **Integration Guide**. Follow the steps to generate a new API key using SendGrid's Web API and cURL.

Reviewer: ```suggestion
3. Create a key using SendGrid's web API. In SendGrid, go to **Email API** > **Integration Guide**. Follow the steps to generate a new API key using SendGrid's Web API and cURL.
```

### Integrate Amazon SES with Astronomer

This setup requires an AWS account and use of the [AWS Management Console](https://aws.amazon.com/console/).

1. In the AWS Management Console, go to **AWS Console** > **Simple Email Service** > **Email Addresses** to add and verify the email addresses you want to receive alerts.

2. Open the inbox of each email address you specified and verify them through the emails sent by Amazon.

3. In the AWS Console, go to **Simple Email Service** > **SMTP Settings** and use the **Create My SMTP Credentials** button to generate a username and password. This will look similar to an access and secret access key. Write down this username and password for step 5, as well as the **Server Name** and **Port**.

   > **Note:** You won't be able to access these values again, so consider storing them in a password manager.

4. Choose an Amazon EC2 region to use, then write down the code of this server for the next step. Refer to [Amazon's list of available regions and servers](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-regions) to determine which server best fits your needs.

> 5. Open your Airflow Deployment in the Astronomer UI and go to the **Variables** tab, then add the following Environment Variables using the **+Add** button:

Reviewer: ```suggestion
5. Open your Airflow Deployment in the Astronomer UI and go to the **Variables** tab. Then, add the following Environment Variables using the **+Add** button:
```

Author: Reworded this slightly

Reviewer: ^Hm sentence still sounds run-on-y to me, so I'd still recommend this change. No big deal either way.

| `AirflowPodQuota` | Your Airflow Deployment has been using over 95% of its pod quota for over 10 minutes. | Either increase your Deployment's Extra Capacity in the Astronomer UI or update your DAGs to use less resources. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |
| `AirflowSchedulerUnhealthy` | The Airflow Scheduler has not emitted a heartbeat for over 1 minute. | Reach out to [Astronomer Support](https://support.astronomer.io). |
| `AirflowTasksPendingIncreasing` | Your Airflow Deployment created tasks faster than it was clearing them for over 30 minutes. | Ensure that your tasks are running and completing correctly. If your tasks are running as expected, [raise concurrency and parallelism in Airflow](https://www.astronomer.io/guides/airflow-scaling-workers), then consider increasing one of the following resources to handle the increase in performance: <ul><li>Kubernetes: Extra Capacity</li><li>Celery: Worker Count or Worker Resources</li></ul><ul><li>Local Executor: Scheduler Resources</li></ul>|

> | `ContainerMemoryNearTheLimitInDeployment` | A container in your Airflow Deployment is near its memory quota; it's been using over 95% of its memory quota for over 60 minutes. | Either increase your Deployment's allocated resources in the Astronomer UI or update your DAGs to use less memory. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |

Author: This content was moved into the Platform Alerts guide, since the general configuration/ concepts are the same. At the beginning of this guide, there's a distinction between Airflow alerts and Deployment/platform alerts and a link to the alternate guide.

> This guide focuses on configuring task and DAG-level alerts. For information on configuring platform and >Deployment alerts, read [Alerting in Astronomer Enterprise]
>(https://www.astronomer.io/docs/enterprise/v0.23/monitor/platform-alerts).

> | `ContainerMemoryNearTheLimitInDeployment` | A container in your Airflow Deployment is near its memory quota; it's been using over 95% of its memory quota for over 60 minutes. | Either increase your Deployment's allocated resources in the Astronomer UI or update your DAGs to use less memory. If you have not already done so, upgrade to [Airflow 2.0](https://www.astronomer.io/blog/introducing-airflow-2-0) for improved resource management. |

Author: I should also note that this won't be the case in the Cloud guides, as they can't configure these alerts in the same way. For that guide, I'll bring information about the Deployment alerts back into the Airflow Alerts guide.
